### PR TITLE
Additional Unit Tests (MATLAB)

### LINKED ISSUE #308: Additional Unit Tests for MATLAB source code
Hi all,

In an effort to make the unit tests more robust, I would like to propose adding some more test cases for the library, starting on the MATLAB side. 

### Operators

Currently, the test suite for each mimetic operator only checks for the row sum property of each operator matrix. As an additional test, I believe for a $k$th order operator in $N$ dimensions ($N=1,2,3$), we should expect that when each operator (```div```, ```grad```, and ```lap```) is applied to a polynomial function of the form

$\sum_{i=1}^{N} x_i^k$

then the numerical calculation matches exactly the analytical calculation. So we can test each operator on a polynomial function sampled on an appropriate grid, and compare with the analytical derivatives of the polynomial function (again, on the appropriate grid points).

### Boundary conditions

Additionally, I would like to extend the unit tests for the boundary conditions, specifically for the addScalarBC functions (in 1D, 2D, and 3D). Using the test cases from julia/MOLE.jl/test/BCs/scalarBC.jl as a reference, I think it would be good to check when no boundary conditions are given, when one of Dirichlet/Neumann are given, and then when both are given. This would give 4 test cases for each dimension of operator.

### Summary

In summary, I would like to add the following:

- 12 new test cases for the divergence (4 polynomial examples in each of 1D, 2D, and 3D)

- 12 new test cases for the gradient (4 polynomial examples in each of 1D, 2D, and 3D)

- 12 new test cases for the laplacian (4 polynomial examples in each of 1D, 2D, and 3D)

- 4 new test cases for each of ```addScalarBC1D```, ```addScalarBC2D```, and ```addScalarBC3D```

Total new test cases: 48

Please let me know if there are any questions or concerns about this proposed addition. Thank you!

### PR DESCRIPTION
**Pull request by mathgeek31415**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Example
- [ ] Documentation

## Description

Additional unit tests to check for exactness of mimetic operators when applied to polynomial functions, and expanded unit tests for boundary operators.

## Related Issues & Documents

- [x] I have created an Issue that is paired with this PR
- Closes #308 

## QA Instructions, Screenshots, Recordings

These tests can be run in MATLAB by opening each file in the editor and clicking "Run Tests". Additionally, the following command should at least work in the MATLAB command window (not sure about Octave):

```runtests(["testDivPolynomial.m" "testGradPolynomial.m" "testLapPolynomial.m" "testAddScalarBC.m"])```

## Added/updated tests?
_We encourage you to test all code included with MOLE, including examples.

- [ ] Yes
- [x] No, and this is why: Technically, these are tests themselves.
- [ ] I need help with writing tests

## Read Contributing Guide and Code of Conduct

- [x] 📖 I have read the MOLE Contributing Guide: https://github.com/csrc-sdsu/mole/blob/main/CONTRIBUTING.md
- [x] 📖 I have read the MOLE Code of Conduct: https://github.com/csrc-sdsu/mole/blob/main/CODE_OF_CONDUCT.md

## [optional] Are there any post deployment tasks we need to perform?

I'm not sure if anything else needs to be done in order to add these to the default test suite along with the existing test files? Any guidance here (if this is something that I should take care of) would be helpful!

### CODE REVIEW

**Comment by mathgeek31415** [comment]
Putting a draft PR here. The new 1D test cases have been committed, and the 2D/3D test cases will be completed soon. The last thing I need to look at before review is the boundary operators.

**Comment by mathgeek31415** [comment]
This PR is now ready for review. In putting together these unit tests, specifically for the addScalarBC functions, I've slightly adjusted what I originally proposed in Issue #308. I found it was redundant to test the Dirichlet and Neumann conditions separately, so I only tested a sample Robin condition instead. I believe this is sufficient as a minimal test case for the boundary operators. This gives a total of 6 unit tests for the boundary operators, instead of 12 unit tests. However, if everyone thinks it would be better to also write tests for just Dirichlet and just Neumann conditions, then I am happy to add this in.

Please let me know if there are any questions, or if there are any issues running the tests.

File: tests/matlab_octave/testAddScalarBC.m [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+            testCase.verifyEqual(A0, A, ...
+                "Matrix A with no boundary conditions doesn't match expected result")
+            testCase.verifyEqual(b0, b, ...
+                "Vector b with no boundary conditions doesn't match expected result")
+
+            % Second test case: with boundary conditions
+
+            dc1 = [2; 3];
+            nc1 = [4; 5];
```
Comment (jbrzensk) [inline tests/matlab_octave/testAddScalarBC.m:31]: Agree that Robin BC should be sufficient to test all cases.

File: tests/matlab_octave/testDivPolynomial.m [hunk 2]
```diff
@@ -0,0 +1,139 @@
+classdef testDivPolynomial < matlab.unittest.TestCase
+    methods (Test)
+        function testDivOfPolynomial1D(testCase)
+            addpath('../../src/matlab_octave')
+            
+            ks = [2, 4, 6, 8];  % Different orders of accuracy
+            tol = 1e-10;
```
Comment (jbrzensk) [inline tests/matlab_octave/testDivPolynomial.m]: Please use 1e-11 for the tol for div and grad. Leave Lap at 1e-10 for the compounded errors.

File: tests/matlab_octave/testDivPolynomial.m [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                sfield = [x1, x1+dx/2:dx:x2-dx/2, x2]';
+                vfield = (x1:dx:x2)';
+
+                approx = D * vfield.^k;
+                analytic = k * sfield.^(k-1);
+                
+                % Note: ignore boundary points for divergence
+                err = norm(approx(2:end-1)-analytic(2:end-1));
+
+                testCase.verifyLessThan(norm(err), tol, ...
```
Comment (jbrzensk) [inline tests/matlab_octave/testDivPolynomial.m:26]: nit: You take the norm on line 24, then again in the test. norm(err) is just extra computations. It is in every test. In all three polynomial files.

**Review by jbrzensk — CHANGES_REQUESTED** [review]
Thanks for adding these. Only notes are to change the err tolerance to 1e-11 for div and grad test cases.

In all of the tests, you take the norm(err) and compare it, but on the previous lines you define err = norm( approx - analytic ). It is unnecessary.

Everything else looks cool.

**Comment by mathgeek31415** [comment]
Hi @jbrzensk , thank you for the feedback! I've removed the duplicate norm calculations and changed the error tolerances in the Div and Grad test files, per your comments. Please let me know if there is anything else that needs to be updated.

**Review by jbrzensk — APPROVED** [review]
